### ISSUE TITLE
Jsoup 1.14.3 & 1.15.3 -> org.jsoup.nodes.Element : Self closing tags must not have children but Jsoup allows them to have children.

### ISSUE THREAD

**Issue by FanJups** [jhy/jsoup#1838]
Hi @jhy ,
Thanks for reviewing this issue.
Sorry, I changed the title because it was confusing. 
Yes, HTML5 does not require a trailing / to make an empty tag. 
I was trying to explain the problem we faced with Jsoup API.

**Our goal is not to parse HTML to get a parsed HTML as output  as it is done here https://try.jsoup.org/ but we are using Jsoup to turn HTML elements into Java Objects (  [Node](https://github.com/jhy/jsoup/blob/master/src/main/java/org/jsoup/nodes/Node.java), [Element](https://github.com/jhy/jsoup/blob/master/src/main/java/org/jsoup/nodes/Element.java) and [TextNode](https://github.com/jhy/jsoup/blob/master/src/main/java/org/jsoup/nodes/TextNode.java) from Jsoup ) then our app turns these Java Objects  into JavaScript variables.**

**While working, we noticed that if there are 2 self closing tags following each other then the 2nd one is considered as the child of the 1st one. To be more precise, we also noticed that when the 1st one has slash at the end then everything is correct but when the 1st one doesn't have slash, we got this issue. After so many trials, we concluded that every tag following a self closing tag without slash is considered as its child.** 

 **It is impossible to see that here https://try.jsoup.org/ . The only way to see that is to deeply analyze the code to get a better idea on how things work under the hood.**

So, how  to reproduce this issue ?

Just run the code below, we invite you to use the 3 strings (selfClosingTagsWorking, selfClosingTagNotWorking and brDiv) to see by yourself. Also run with many examples as you want to better understand.

```java
import org.jsoup.Jsoup;
import org.jsoup.nodes.Element;
import org.jsoup.parser.Parser;

public class SelfClosingTagIssue {

    public static void main(String[] args) {

	String selfClosingTagWorking = "<input type=\"text\"/>\r\n" + "<img src=\"#URL\" alt=\"image\">";

	String selfClosingTagNotWorking = "<input type=\"text\">\r\n" + "<img src=\"#URL\" alt=\"image\">";

	String brDiv = "<br> <div><div/>";

	Element htmlDoc = Jsoup.parse(brDiv, "", Parser.xmlParser());

	for (Element child : htmlDoc.children()) {

	    boolean hasChild = child.childrenSize() > 0;

	    System.err.println(" **** Analyze :" + child.tag().getName() + " -> isEmpty : " + child.tag().isEmpty()
		    + " -> isSelfClosing : " + child.tag().isSelfClosing() + " -> hasChild : " + hasChild + " **** ");

	    if (hasChild) {
		System.err.println("Children of " + child.tagName() + ": ");

		for (Element e : child.children()) {

		    System.err.println(e.tagName());

		}
	    }

	}

    }

}

```

Result with selfClosingTagWorking

```
 **** Analyze :input -> isEmpty : true -> isSelfClosing : true -> hasChild : false **** 
 **** Analyze :img -> isEmpty : true -> isSelfClosing : true -> hasChild : false **** 

```

Result from selfClosingTagNotWorking

```
 **** Analyze :input -> isEmpty : true -> isSelfClosing : true -> hasChild : true **** 
Children of input: 
img

```

Result from brDiv

```
 **** Analyze :br -> isEmpty : true -> isSelfClosing : true -> hasChild : true **** 
Children of br: 
div

```

_Originally posted by @FanJups in https://github.com/jhy/jsoup/issues/1741#issuecomment-1246952249_

**Comment by SalathielGenese** [comment]
Just to say... In normal HTML:
```html
<div>
  <p>lorem ipsum...
```

Is interpreted as:
```html
<div>
  <p>lorem ipsum...</p></div>
```

At least, that is how browsers process these.

I was able to find WHATWG reference about even a more strange behaviour: [13.2.10.6 Unclosed formatting elements](https://html.spec.whatwg.org/multipage/parsing.html#unclosed-formatting-elements).

---
And there is more complications to those rules: there is an entry **A start tag whose tag name is "a"** at [13.2.6.4.7 The "in body" insertion mode](https://html.spec.whatwg.org/multipage/parsing.html#parsing-main-inbody) which reads

> In the non-conforming stream `<a href="a">a<table><a href="b">b</table>x`, the first [a](https://html.spec.whatwg.org/multipage/text-level-semantics.html#the-a-element) element would be closed upon seeing the second one, [a](https://html.spec.whatwg.org/multipage/text-level-semantics.html#the-a-element)nd the "x" character would be inside a link to "b", not to "a". This is despite the fact that the outer a element is not in table scope (meaning that a regular `</a>` end t[a](https://html.spec.whatwg.org/multipage/text-level-semantics.html#the-a-element)g [a](https://html.spec.whatwg.org/multipage/text-level-semantics.html#the-a-element)t the start of the table wouldn't close the outer a element). The result is that the two a elements are indirectly nested inside each other — non-conforming markup will often result in non-conforming DOMs when parsed.

---
It is interesting not to find `input` element listed under [13.2.6.3 Closing elements that have implied end tags](https://html.spec.whatwg.org/multipage/parsing.html#closing-elements-that-have-implied-end-tags)

**Comment by jhy** [comment]
You are parsing with the XML parser. It does not know anything about HTML because it is not a HTML parser. So it treats `<img>` as an opening tag, not a void tag.

I don't see any issue here.

**Comment by FanJups** [comment]
Thanks for helping us to figure it out. **There is an issue (based on our goal not jsoup issue) when using xmlParser. On the other hand, hmlParser has no issue. We use xml instead of  html parser because html parser adds html,head and body tags but we don't want that. We used xml then added a logic to get a similar result we got using htmlParser but without html,head and body tags.**

There are 2 branches with examples  in order to better understand:


xmlParser:  https://github.com/osscameroon/js-generator/tree/basic-illustration-of-self-closing-tag-issue

htmlParser:  https://github.com/osscameroon/js-generator/tree/basic-illustration-of-self-closing-tag-issue-htmlparser
